File: 2Darray.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int row = rowSearch(matrix,target);
        
        if (row == -1)
        {
            return false;
        }
        
        boolean isFound = binarySearch(matrix, row, target);
        
        return isFound;
    }
    
    public static int rowSearch(int [][] matrix, int target)
    {
        int low = 0;
        int high = matrix.length - 1;
        int lastColoumn = matrix[0].length - 1;
        
        while( low <= high)
        {
           int mid = low + (high - low)/2;
            
            if(target >= matrix[mid][0]   && target <= matrix[mid][lastColoumn])
            {
                return mid;
            }
            else if (target > matrix[mid][0])
            {
                low = mid + 1;
            }
            else if (target < matrix[mid][0])
            {
                high = mid - 1;
            }
        }
    return -1;
    }
    public static boolean binarySearch(int [][] matrix, int r, int target)
    {
        int low = 0;
        int high = matrix[0].length - 1;
        
        while (low <=high)
        {
            int mid = low + (high - low)/2;
            
            if(matrix[r][mid] == target)
            {
                return true;
            }
            else if (matrix[r][mid] < target)
            {
                low = mid + 1;
            }
            
            else if(matrix[r][mid] > target)
            {
                high = mid - 1;
            }
                
        }
        return false;
    }
    
}

File: 2darray.py
'''
Time Complexity: O(logn)
Space Complexity: O(1)
'''
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row = len(matrix)
        col = len(matrix[0])
        low = 0
        high = (row*col)-1
        while(low<=high):
            mid = int((low+high)/2)
            val = matrix[int(mid/col)][mid%col]
            if(val==target):
                return True
            if(val<target):
                low = mid+1
            if(val>target):
                high = mid-1
        return False